posts/building-a-speech-enabled-application-using-twilio-and-aws-with-jhipster.md

```diff
@@ -0,0 +1,151 @@
+    }
+
+    // Generate TwiML
+    Say say = new Say.Builder(voiceCall.getMessage()).voice(Say.Voice.valueOf("POLLY_" + voiceCall.getVoice().name())).build();
+    VoiceResponse twiml = new VoiceResponse.Builder()
+        .say(say)
+        .pause(new Pause.Builder().length(1).build())
+        .build();
+
+    // Update VoiceCall
+    voiceCall.twiml(twiml.toXml())
+        .date(ZonedDateTime.now());
+
+    VoiceCall result = voiceCallRepository.save(voiceCall);
+
+    // Save TwiML to S3 with public read
+    PutObjectRequest putRequest = PutObjectRequest.builder()
+        .bucket(BUCKET)
+        .key(result.getId() + ".xml")
+        .contentType("text/xml")
+        .acl(ObjectCannedACL.PUBLIC_READ)
+        .build();
+    s3.putObject(putRequest, software.amazon.awssdk.core.sync.RequestBody.fromString(result.getTwiml()));
+
+    // Make call using Twilio
+    String twimlUri = "https://s3.amazonaws.com/" + putRequest.bucket() + "/" + putRequest.key();
+    Call call = Call.creator(new PhoneNumber(voiceCall.getNumber()), new PhoneNumber(TWILIO_NUMBER), new URI(twimlUri))
+        .setMethod(HttpMethod.GET)
+        .create();
+    log.debug("CALL SID : {}", call.getSid());
+
+    return ResponseEntity.created(new URI("/api/voice-calls/" + result.getId()))
+        .headers(HeaderUtil.createEntityCreationAlert(ENTITY_NAME, result.getId().toString()))
+        .body(result);
+}
+```
+
+Here is the workflow:
+1. Generate a TwiML with user's message and voice
+2. Update VoiceCall with TwiML and current date
+3. Save TwiML to S3
+4. Call user's number with Twilio using the TwiML from S3
+5. Return created VoiceCall with a HTTP 201
+
+To keep this post simple, I decided to upload the TwiML to S3 with a public read access. I recommend using a [presigned URL](https://docs.aws.amazon.com/AmazonS3/latest/dev/ShareObjectPreSignedURL.html), especially for an application going to production.
+
+## Testing
+
+To test that everything is working fine, the application must be first started with the command `./mvnw`. You can then browse to [http://localhost:8080/#/entity/voice-call/new](http://localhost:8080/#/entity/voice-call/new) to create a VoiceCall (don't forget the `+1` in front of your number).
+
+![](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2019/03/twilio.png)
+
+If you don't receive the call after saving, the [Twilio's calls log](https://www.twilio.com/console/voice/calls/logs) keeps track of all the calls and can be used for troubleshooting.
+
+# Conclusion
+
+Building a speech-enabled application might sound difficult but with the help of Twilio and AWS it is actually a very easy duty. The developer experience is also very good since Twilio provides a SDK in 6 languages (Node, C#, PHP, Ruby, Python and Java) but they also provide a [REST API](https://www.twilio.com/docs/usage/api). Amazon Polly gives you a large panel of realistic voices to make your application fully international by supporting multiple languages.
```

Comment:
```suggestion
Building a speech-enabled application might sound difficult but with the help of Twilio and AWS it is actually very easy. The developer experience is also very good since Twilio provides a SDK in 6 languages (Node, C#, PHP, Ruby, Python and Java) and they also provide a [REST API](https://www.twilio.com/docs/usage/api). Amazon Polly gives you a large panel of realistic voices to make your application fully international by supporting multiple languages.
```